src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -56,9 +56,33 @@ def get_preference(
         information  # type: Sequence[Tuple[Requirement, Candidate]]
     ):
         # type: (...) -> Any
+        """Produce a sort key for given requirement based on preference.
+
+        The lower the return value is, the more preferred this group of
+        arguments is.
+
+        Currently pip considers the followings in order:
+
+        * Prefer if any of the known requirements points to an explicit URL.
+        * If equal, prefer if any requirements contain `===` and `==`.
+        * If equal, prefer user-specified (non-transitive) requirements.
+        * If equal, order alphabetically for consistency (helps debuggability).
+        """
+
+        def _get_restrictive_rating(requirements):
+            # type: (Iterable[Requirement]) -> int
+            lookups = (r.get_candidate_lookup() for r in requirements)
+            cands, ireqs = zip(*lookups)
+            if any(cand is not None for cand in cands):
+                return 0
+            if any(ireq.is_pinned for ireq in ireqs if ireq is not None):
+                return 1
+            return 2
```

Comment:
We should probably also differentiate between `pkg<version` and plain `pkg`?

Reply:
Maybe, and I think `>` and `>=` may be even more useful since they generally only show up when the package author is really sure about the version range. I thought about all of the operators briefly but decided it would take us too deep into the rabbit hole. Let’s get this in first and discuss further optimisations later.

Reply:
Now that we've broken out the operator comparisions... I feel like it's a matter of adding a check for:

<=, >=: Inclusive ordered comparison clause
<, >: Exclusive ordered comparison clause
 
It's easy enough to add as a "2", and bumping the final return to 3. If we're all on the same page on that, let's update this PR and mark this conversation as resolved on the PR. If not, let's file an issue and mark this conversation as resolved on the PR. :P

Reply:
See latest commit.